R/pkg-versions.R

```diff
@@ -0,0 +1,256 @@
+# GitHub releases.
+
+check_pkg_version_path <- function () {
+    cm_path <- normalizePath (file.path ("raw_cm.json"), mustWork = FALSE)
+
+    if (!file.exists (cm_path)) {
+        stop (
+            "This function can only be run in the ",
+            "output directory of 'roregistry', which ",
+            "must contain the 'raw_cm.json' file.",
+            call. = FALSE
+        )
+    }
+
+    return (cm_path)
+}
+
+#' Extract data on releases of packages, both on CRAN and GitHub.
+#'
+#' @param CRAN_only If `TRUE`, return only those rows of `repo_data` which
+#' correspond to packages which are or were on CRAN; otherwise return all data.
+#' @return A `data.frame` with one row for each package, and columns containing
+#' information on the package including dates and version numbers of GitHub and
+#' CRAN releases (if any).
+#'
+#' @export
+registry_pkg_versions <- function (CRAN_only = TRUE) {
+
+    cm_path <- check_pkg_version_path ()
+
+    # Code from list-packages.R to get all pkgs:
+    hosted_packages <- get_hosted_packages()
+    other_packages <- get_other_packages()
+    packages <- c(hosted_packages, other_packages)
+    packages <- packages[order(purrr::map_chr(packages, "package"))]
+    packages <- lapply (packages, function (i) i [c ("package", "url", "branch")])
+    packages <- data.frame (do.call (rbind, lapply (packages, unlist)))
+
+    ptn <- "^https\\:\\/\\/github\\.com\\/|\\/.*$"
+    packages$org <- gsub (ptn, "", packages$url)
+    packages$repo <- gsub ("^.*\\/", "", packages$url)
+    # All "official" packages are in "ropensci" only:
+    index <- which (grepl ("ropensci|r-lib", packages$org) & packages$org != "ropensci")
+    packages <- packages [-index, ]
+    packages_other <- packages [which (!packages$org == "ropensci"), c ("org", "repo")]
+
+    packages_other_dat <- apply (packages_other, 1, function (i) {
+        get_pkg_release_data (org = as.character (i [1]), repo = as.character (i [2]))
+    })
+    packages_other_dat <- do.call (rbind, packages_other_dat)
+
+    rbind (get_pkg_releases_data ("ropensci"), packages_other_dat) |>
+        add_pkg_name (cm_path) |>
+        add_CRAN_version (CRAN_only = CRAN_only)
+}
+
+#' Main function to extract and collate data on package releases, both on
+#' GitHub and CRAN.
+#'
+#' @param org Name of GitHub organization for which data are to be extracted.
+#' @param n Number of entries returned from a single queries. Queries are run
+#' until all data are extracted, so this parameter has no effect, and should
+#' not be modified.
+#' @return A `data.frame` of package data.
+#'
+#' @noRd
+get_pkg_releases_data <- function (org = "ropensci", n = 100L) {
+
+    q <- get_releases_query (org = org, n = n)
+    dat <- gh::gh_gql (query = q)
+    page_info <- dat$data$repositoryOwner$repositories$pageInfo
+    repo_data <- dat$data$repositoryOwner$repositories$nodes
+    while (page_info$hasNextPage) {
```

Comment:
why get all releases and not just the latest one that's not a draft? https://github.com/ropensci/roweb3/blob/e4042ee863c9e0992ff3d34b1d81b9c9792b176d/archetypes/newsletter/index.md?plain=1#L228

Reply:
and why use the V4 API for this?

Reply:
All releases because at least initially we need all that information that we currently don't have. And v4 API because we can grab the whole org in one call; with the REST api, we'd have to do each repo as a separate call.